src/map/npc.cpp

```diff
@@ -900,16 +913,28 @@ int npc_event(struct map_session_data* sd, const char* eventname, int ontouch)
 
 	switch(ontouch)
 	{
-	case 1:
-		nd->touching_id = sd->bl.id;
-		sd->touching_id = nd->bl.id;
-		break;
-	case 2:
-		sd->areanpc_id = nd->bl.id;
-		break;
-	}
+	case 1: {
+			if (pc_ishiding(sd))
+				return 0;
 
-	return npc_event_sub(sd,ev,eventname);
+			nd->touching_id = sd->bl.id;
+
+			auto it = std::find(sd->npc_ontouch_.begin(), sd->npc_ontouch_.end(), nd->bl.id);
+
+			if (it == sd->npc_ontouch_.end())
+				sd->npc_ontouch_.push_back(nd->bl.id);
+			break;
+		}
+	case 2: {
+			auto it = std::find(sd->areanpc.begin(), sd->areanpc.end(), nd->bl.id);
+
+			if (it == sd->areanpc.end())
+				sd->areanpc.push_back(nd->bl.id);
+			break;
+		}
+	}
+	npc_event_sub(sd,ev,eventname);
```

Comment:
You are discarding the return value of `npc_event_sub`. I would like to suggest that you don't use switch statement here since it's just only two cases. It look much cleaner.
```suggestion
if (ontouch == 1) {
    if (pc_ishiding(sd))
        return 0;

    nd->touching_id = sd->bl.id;

    auto it = std::find(sd->npc_ontouch_.begin(), sd->npc_ontouch_.end(), nd->bl.id);
    if (it == sd->npc_ontouch_.end())
        sd->npc_ontouch_.push_back(nd->bl.id);

} else if (ontouch == 2) {
    auto it = std::find(sd->areanpc.begin(), sd->areanpc.end(), nd->bl.id);
    if (it == sd->areanpc.end())
        sd->areanpc.push_back(nd->bl.id);
}

return npc_event_sub(sd,ev,eventname);
```